source/h2_connection.c

```diff
@@ -256,16 +272,326 @@ static void s_handler_destroy(struct aws_channel_handler *handler) {
+    struct aws_linked_list *outgoing_frames_queue = &connection->thread_data.outgoing_frames_queue;
+    struct aws_linked_list *outgoing_streams_list = &connection->thread_data.outgoing_streams_list;
+
+    AWS_PRECONDITION(aws_channel_thread_is_callers_thread(channel_slot->channel));
+    AWS_PRECONDITION(connection->thread_data.is_outgoing_frames_task_active);
+
+    /* If there is nothing to send, then end the task immediately */
+    if (aws_linked_list_empty(outgoing_frames_queue) && aws_linked_list_empty(outgoing_streams_list)) {
+        CONNECTION_LOG(TRACE, connection, "Outgoing frames task stopped, nothing to send at this time");
+        connection->thread_data.is_outgoing_frames_task_active = false;
+        return;
+    }
+
+    /* Acquire aws_io_message, that we will attempt to fill up */
+    struct aws_io_message *msg = aws_channel_slot_acquire_max_message_for_write(channel_slot);
+    if (AWS_UNLIKELY(!msg)) {
+        CONNECTION_LOG(ERROR, connection, "Failed to acquire message from pool, closing connection.");
+        goto error;
+    }
+
+    /* Set up callback so we can send another message when this one completes */
+    msg->on_completion = s_on_channel_write_complete;
+    msg->user_data = connection;
+
+    CONNECTION_LOGF(
+        TRACE,
+        connection,
+        "Outgoing frames task acquired message with %zu bytes available",
+        msg->message_data.capacity - msg->message_data.len);
+
+    /* Track number of frames encoded, just used for logging */
+    size_t num_frames_encoded = 0;
+
+    /* Write as many frames from outgoing_frames_queue as possible. */
+    while (!aws_linked_list_empty(outgoing_frames_queue)) {
+        struct aws_linked_list_node *frame_node = aws_linked_list_front(outgoing_frames_queue);
+        struct aws_h2_frame_base *frame = AWS_CONTAINER_OF(frame_node, struct aws_h2_frame_base, node);
+
+        /* #TODO actual functionality to query min required space for a frame */
+        const size_t min_required_bytes = 1024;
+        const size_t available_bytes = msg->message_data.capacity - msg->message_data.len;
+        if (available_bytes < min_required_bytes) {
+            if (msg->message_data.len == 0) {
+                /* We're in trouble if an empty message isn't big enough for this frame to do any work with */
+                CONNECTION_LOGF(
+                    ERROR,
+                    connection,
+                    "Cannot encode %s frame requiring %zu bytes, max available space is %zu",
+                    aws_h2_frame_type_to_str(frame->type),
+                    min_required_bytes,
+                    available_bytes);
+                aws_raise_error(AWS_ERROR_INVALID_STATE);
+                goto error;
+            }
+
+            CONNECTION_LOG(TRACE, connection, "Outgoing frames task filled message, and has more frames to send later");
+            goto done_encoding;
+        }
+
+        /* #TODO some way for frame to say it's not done yet.
```

Comment:
why not track the amount written in the frame and just leave it in the queue until the entire thing is sent? This of course adds some craziness because you now have to make sure no data frames go out until you're done with this part, but that should be fine since you always prefer this queue.

Reply:
there's currently no way to query the frame's total encoded length, so I can't just compare total length against the amount written. You're right though, that with this design I can split any non-data frame across `aws_io_messages` no problem.

I think my next PR will be overhauling this encoder to work the way I need it to.